### PR TITLE
psa: First drop of PSA into Matter

### PR DESCRIPTION
**Pull request by adbridge**
This PR brings PSA cryptography into the ARM-software version of Matter and builds with the SDK.

There is an update to the VSCode docker image to allow the SDK unit tests and examples to be built with PSA or mbedtls via the Tasks menus.

There is also an update to the CI workflow for the SDK examples to build and run the tests with both mbedtls and psa cryptographic algorithms.

### CODE REVIEW

**Comment by ATmobica** [comment]
LGTM, just create the a separate commit with a fix in `crypto.gni` file

**Comment by adbridge** [comment]
> @adbridge LGTM, just create a separate commit with a fix in `crypto.gni` file

I decided to cover it in the same commit but document what was done and why there. This should suffice yes ?

**Review by ATmobica — COMMENTED** [review]
Could you apply restyler on your changes?

**Comment by adbridge** [comment]
> Could you apply restyler on your changes?

Restyler?

File: examples/platform/openiotsdk/app/openiotsdk_platform.cpp [hunk 1]
```diff
@@ -188,6 +189,13 @@ int openiotsdk_platform_init(void)
         return EXIT_FAILURE;
     }
 
+    ret = psa_crypto_init();
```
[the remark this answers is not in the document]
Comment (adbridge) [reply examples/platform/openiotsdk/app/openiotsdk_platform.cpp:192]: I thought about having a define, but it works both with and without psa being available but we can certainly try and guard it anyway.

[the remark this answers is not in the document]
Comment (ATmobica) [reply examples/platform/openiotsdk/app/openiotsdk_platform.cpp:192]: Yeah, it works because we set TF-M support by default to each example. But let's guard it in case that ever changes.

[the remark this answers is not in the document]
Comment (adbridge) [reply examples/platform/openiotsdk/app/openiotsdk_platform.cpp:192]: I would suggest
#if defined(MBEDTLS_PSA_CRYPTO_C)
as this is where similar code is guarded elsewhere and that code is part of the mbedtls PSA still.

[the remark this answers is not in the document]
Comment (pan-) [reply examples/platform/openiotsdk/app/openiotsdk_platform.cpp:192]: Your suggestion looks good to me.

File: scripts/examples/openiotsdk_example.sh [hunk 2]
```diff
@@ -57,8 +58,9 @@ Options:
     -C,--command    <command>           Action to execute <build-run | run | test | build - default>
     -d,--debug      <debug_enable>      Build in debug mode <true | false - default>
     -S,--socket     <iotsocket|lwip>    Build with IoT Socket API (enabled by default)
-    -p,--path       <build_path>        Build path <build_path - default is example_dir/build>
-    -n,--network    <network_name>      FVP network interface name <network_name - default is "user" which means user network mode>
+    -a,--algorithm  <crypto algorithm) Select crypto algorithm <psa | mbedtls - default>
```
Comment (pan-) [inline scripts/examples/openiotsdk_example.sh:61]: crypto backend ?

File: scripts/examples/openiotsdk_example.sh [hunk 3]
```diff
@@ -41,6 +41,7 @@ TELNET_CONNECTION_PORT=""
 FAILED_TESTS=0
 IS_UNIT_TEST=0
 FVP_NETWORK="user"
+CRYPTO_TYPE="mbedtls"
```
Comment (pan-) [inline scripts/examples/openiotsdk_example.sh:44]:
```suggestion
CRYPTO_BACKEND="mbedtls"
```

File: src/crypto/CHIPCryptoPALPSA.cpp [hunk 4]
```diff
@@ -40,6 +40,7 @@
 #include <mbedtls/oid.h>
 #include <mbedtls/x509.h>
 #include <mbedtls/x509_csr.h>
+#include <mbedtls/threading.h>
```
Comment (pan-) [inline src/crypto/CHIPCryptoPALPSA.cpp:43]: Why is this required now ?

**Review by pan- — COMMENTED** [review]
Thanks for the change @adbridge. 

I suggest having a single term to refer to the crypto backend instead of multiple (`type` or  `algorithm` at the moment). 

I think we can name it a `backend` or `implementation`. Whatever is chosen, we should use that name consistently in all our references (scripts, Github workflows, vscode tasks, ...)

File: src/crypto/CHIPCryptoPALPSA.cpp (hunk 4 above)
Comment (adbridge) [reply src/crypto/CHIPCryptoPALPSA.cpp]: It wasn't building previously without this, I'll try it without again and see if it is still required.

**Comment by adbridge** [comment]
@pan- @ATmobica that should be all the review comments done...

File: examples/platform/openiotsdk/app/openiotsdk_platform.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ATmobica) [reply examples/platform/openiotsdk/app/openiotsdk_platform.cpp:196]: I'm not sure whether the dependency on mbedtls configuration is correct at this point. I think about something like that: 
```suggestion
#if (CONFIG_CHIP_CRYPTO == "psa")
    ret = psa_crypto_init();
#endif
```
Then we need to pass the `CONFIG_CHIP_CRYPTO' flag to `openiotsdk-app` target. The same as we did with TF-M support: https://github.com/ARM-software/connectedhomeip/blob/89039726d54eab72526aa66868779fa29d465351/config/openiotsdk/CMakeLists.txt#L288

[the remark this answers is not in the document]
Comment (adbridge) [reply examples/platform/openiotsdk/app/openiotsdk_platform.cpp:196]: I'm not sure I agree at this point. The current implementation of PSA relies on some underlying calls to mbedtls still. Those calls themselves are guarded by MBEDTLS_PSA_CRYPTO_C so that make far more sense to me. If in the future the PSA implementation moves away from its reliance on mbedtls then I think your suggestion then makes more sense. The logic goes like this, we could have CONFIG_CHIP_CRYPTO set to psa but MBEDTLS_PSA_CRYPTO_C set False and it wouldn't work....

[the remark this answers is not in the document]
Comment (ATmobica) [reply examples/platform/openiotsdk/app/openiotsdk_platform.cpp:196]: Sure, but let's look at the other side. If we have `MBEDTLS_PSA_CRYPTO_C` enabled in our mbedtls does it mean that we want to use PSA crypto in the application? If we build an application with mbedtls crypto we still initialize the psa_crypto because the mbedtls_config stays the same.

[the remark this answers is not in the document]
Comment (adbridge) [reply examples/platform/openiotsdk/app/openiotsdk_platform.cpp:196]: Maybe we should have both, nested?
#if (CONFIG_CHIP_CRYPTO == "psa")
#if defined(MBEDTLS_PSA_CRYPTO_C)

??

[the remark this answers is not in the document]
Comment (ATmobica) [reply examples/platform/openiotsdk/app/openiotsdk_platform.cpp:196]: I think `#if (CONFIG_CHIP_CRYPTO == "psa")` should be enough.

[the remark this answers is not in the document]
Comment (ATmobica) [reply examples/platform/openiotsdk/app/openiotsdk_platform.cpp:196]: LGTM

**Comment by adbridge** [comment]
@pan- @ATmobica I've pushed the update for the guards. Tested compiling both with and without the psa flag and also the psa flag whilst temporarily disabling the MBEDTLS PSA CRYPTO flag:
```
_CONFIG_H=1 -I../../../../../../src/include -I../../../../../../src -Igen/include -I../../../../../../zzz_generated/app-common -I../../../../../../third_party/nlassert/repo/include -I../../../../../../third_party/nlio/repo/include -I../../../../../../third_party/nlfaultinjection/repo/include -c ../../../../../../src/crypto/CHIPCryptoPALPSA.cpp -o obj/src/crypto/cryptopal_psa.CHIPCryptoPALPSA.cpp.o
../../../../../../src/crypto/CHIPCryptoPALPSA.cpp:51:2: error: #error "PSA crypto API not enabled in mbedTLS config"
   51 | #error "PSA crypto API not enabled in mbedTLS config"
      |  ^~~~~
cc1plus: note: unrecognized command-line option '-Wno-unknown-warning-option' may have been intended to silence earlier diagnostics

However it looks like the guard in CHIPCryptoPALPSA.cpp is hit before we hit our new guard, so we may not need the new one after all, although won't do any harm. Thoughts ?

File: examples/platform/openiotsdk/app/openiotsdk_platform.cpp [hunk 5]
```diff
@@ -29,6 +29,10 @@
 #include "iotsdk/ip_network_api.h"
 #include "mbedtls/platform.h"
 
+#if defined(MBEDTLS_PSA_CRYPTO_C)
```
Comment (ATmobica) [inline examples/platform/openiotsdk/app/openiotsdk_platform.cpp:32]:
```suggestion
#ifdef CONFIG_CHIP_CRYPTO_PSA
```

File: config/openiotsdk/cmake/chip.cmake [hunk 6]
```diff
@@ -50,16 +51,20 @@ if(TFM_SUPPORT)
     if(CONFIG_CHIP_OPEN_IOT_SDK_USE_PSA_PS)
         add_dependencies(chip-gn tfm-ns-interface)
     endif()
+
+if ("${CONFIG_CHIP_CRYPTO}" STREQUAL "psa")
+    target_compile_definitions(openiotsdk-chip INTERFACE
+        CONFIG_CHIP_CRYPTO_PSA)
```
Comment (ATmobica) [inline config/openiotsdk/cmake/chip.cmake:57]:
```suggestion
    target_compile_definitions(openiotsdk-chip 
        INTERFACE
            CONFIG_CHIP_CRYPTO_PSA
     )
```

**Review by ATmobica — APPROVED** [review]

**Comment by adbridge** [comment]
@pan- @ATmobica the test failures look unrelated to the changes! I've also spent this afternoon running the unit tests locally (the failure shown for the lwip/mbedtls configuration) as well as with psa and they all pass for me!!

**Review by pan- — APPROVED** [review]

**Comment by pan-** [comment]
One rebase and we can merge @adbridge

**Comment by adbridge** [comment]
> One rebase and we can merge @adbridge

Looks like Artur already dropped his commit onto this branch so I think it is good to merge
